[Tooling] Bump golang to 1.20

okdas opened the pull request:
## Description

After the discussion about WASM and Go 1.21, I thought we should upgrade to 1.20 first to make sure the 1.19 -> 1.20 upgrade can be done without issues, for easier upgrade to 1.21 later.

We have a ticket #588 that touches on that, but it also recommends another change that can be taken care of in another PR later.

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 20 Jul 23 18:54 UTC
This pull request includes several changes to bump the golang version to 1.20. The patch modifies various files such as workflows, Dockerfiles, and documentation files. Additionally, a new document called GOLANG_UPGRADE.md is added to provide a checklist for future Go version upgrades.
<!-- reviewpad:summarize:end -->

#508

Fixes #<issue_number>

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [x] Compiler version change

## List of changes

- Bumped Go to 1.20

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [x] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

dylanlott reviewed (changes requested):
The `go.mod` and `build.mk` files still reference go 1.18, and I believe they need to be updated to 1.20 as well.

dylanlott commented:
One more I found: `docs/development/README.md` references go 1.18 in a comment, I think that should be updated, too.

dylanlott commented on .github/workflows/golangci-lint.yml:
```diff
@@ -10,7 +10,7 @@ permissions:
 env:
   # Even though we can test against multiple versions, this one is considered a target version.
-  TARGET_GOLANG_VERSION: "1.19"
+  TARGET_GOLANG_VERSION: "1.20"
```
👍

h5law commented:
memory arenas?

Olshansk commented:
> LGTM ✅ I was able to boot LocalNet, get a local shell, and run `make test_e2e` with passing on this so I think we're good to go. I'm going to approve this, let's get up to 1.21 soon ✨ I also put together a [small guide](https://www.notion.so/pocketnetwork/Pocket-Go-Version-Upgrades-5585114c358d4bb182b0c27a4c219f59?pvs=4) for updating Go versions in Pocket since we will likely be doing this sort of upgrade with some regularity.

Discussed offline but posting here for posterity - let's move the guide (and link this as an example PR) in the source code.

Make sure to also message the whole team with an announcement (e.g. golang upgrade) and instructions (e.g. wipe docker containers) once it's merged.

okdas commented:
Thank you, @dylanlott, for that document, I've moved it to this repo. 

@Olshansk I also changed the Go version in the main README, seems like the new shields had 1.10 mentioned 🦕
